### PR TITLE
fix(picks): atomically remove ownerIds entry in unjoinOption

### LINKED ISSUE #128: fix(options): unjoinOption read-then-write has the same race as removeMember (#108)
Spotted while reviewing PR #97. The new `unjoinOption(pickId, optionId, ownerUid)` in `src/server/data/options.ts` does a `get()` to fetch `ownerIds`, filters out `ownerUid` in memory, and then either deletes the option (if `remaining` is empty) or removes the single uid via `ownersRef.child(ownerUid).remove()`. This is the same read-then-write pattern that bit `removeMember` (tracked in #108).

The benign races are fine; the problem case is two concurrent unjoin calls on an option with two owners: both reads see `[A, B]`, both decide `remaining > 0` (so don't delete), both individually remove their own uid → `ownerIds` ends up empty but the option still exists. The option is now owner-less, which contradicts the PR's invariant.

A Firebase RTDB transaction on `picks/${pickId}/options/${optionId}/ownerIds` (or on the option node itself) would resolve atomically. Pre-launch this is a theoretical concern; a fix can mirror whatever pattern lands on #108.

---
*Created by Claude Sonnet 4.6*

### PR DESCRIPTION
**Pull request by rmartz**
Replace the read-then-write pattern in `unjoinOption` with a Firebase RTDB transaction on `ownerIds`, mirroring the fix applied to `removeMember` in #129.

The old implementation did a `get()` to fetch `ownerIds`, filtered in memory, then either deleted the option or removed the single uid. Two concurrent unjoin calls on a two-owner option could both read `[A, B]`, both decide "remaining > 0 so don't delete", and both remove only their own uid — leaving the option alive with an empty `ownerIds` map. The transaction aborts (returns `undefined`) when the calling uid is the only owner, triggering a follow-up delete of the whole option node; otherwise it removes the uid atomically in a single write.

## Acceptance Criteria
- [x] `unjoinOption` uses a Firebase RTDB transaction (not a pre-read) on `ownerIds`
- [x] Transaction update function aborts when the calling uid is the only owner
- [x] Transaction update function returns `ownerIds` minus the calling uid when other owners remain
- [x] Option node is deleted and `{ deleted: true }` returned when the transaction does not commit (last-owner case)
- [x] Option node is not deleted and `{ deleted: false }` returned when the transaction commits

## Tests
11 tests added, all passing.

Closes #128

---
*Created by Claude Sonnet 4.6*

### CODE REVIEW

**Review by rmartz — COMMENTED** [review]
## Verdict

All acceptance criteria met, CI passing, no issues found — ready to merge after UAT.

---

## Overview

Replaces the `get()`-then-write pattern in `unjoinOption` with a Firebase RTDB transaction on `ownerIds`. The transaction update function aborts (returns `undefined`) when the calling uid is the last owner, triggering a follow-up delete of the whole option node; otherwise it returns `ownerIds` minus the uid in a single atomic write. This exactly mirrors the fix applied to `removeMember` in #129.

## Code quality

The implementation is clean and correct. The `if (!owners) return owners` arm handles the null (non-existent node) case by passing it through — Firebase RTDB treats a null return as an abort, so the follow-up delete path fires correctly. The `length <= 1` guard covers the single-owner case and any degenerate empty-map case. `Object.fromEntries(Object.entries(...).filter(...))` is an idiomatic and type-safe way to remove a key without mutating the input.

## Correctness

The race condition described in #128 is fully closed: since the uid removal and the "any other owners?" decision happen inside a single transaction, two concurrent unjoin calls on a two-owner option can no longer both see the other owner present and both skip deletion.

One subtle nuance worth noting: when `owners` is null and the transaction aborts, the follow-up `remove()` call is a no-op on an already-absent node — harmless. And when `owners` is non-null but the uid being removed is not actually present in the map (e.g. a double-unjoin), the transaction still commits (the uid just isn't there to filter out), `remove()` is skipped, and `{ deleted: false }` is returned. That behaviour is consistent with the old implementation and appropriate.

## Test coverage

9 tests added, structured one-assertion-per-test against each acceptance criterion. The test for the null snapshot case (criterion 2, second `it`) correctly verifies that `capturedUpdate?.(null)` returns `null` rather than `undefined` — ensuring the abort path only fires for the single-owner case, not for missing data.

---

*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

The following items were raised in Copilot's review (posted after the previous human review) and remain unaddressed:

- ❌ [src/server/data/options.ts line 64 (copilot-pull-request-reviewer)](https://github.com/rmartz/group-picks/pull/153#discussion_r3238814362): Non-owner uid can trigger option deletion via the `length <= 1` abort check.
- ❌ [src/server/data/options.ts line 63 (copilot-pull-request-reviewer)](https://github.com/rmartz/group-picks/pull/153#discussion_r3238814393): `return owners` when `owners === null` commits rather than aborts the RTDB transaction.
- ❌ [src/server/data/options.spec.ts line 95 (copilot-pull-request-reviewer)](https://github.com/rmartz/group-picks/pull/153#discussion_r3238814407): Test incorrectly asserts that `null` is returned for the null-owners case, but `null` commits rather than aborts.
- ❌ [src/server/data/options.spec.ts line 80 (copilot-pull-request-reviewer)](https://github.com/rmartz/group-picks/pull/153#discussion_r3238814421): No test for the case where `ownerUid` is not present in `ownerIds`.

---

## Requested changes

- **Non-owner can delete an option (privilege escalation bug)**: The transaction update function aborts (returns `undefined`) whenever `Object.keys(owners).length <= 1`, regardless of whether the calling uid is actually in the owners map. If a user who is NOT an owner calls `unjoinOption` on an option with exactly one owner, the transaction aborts, the follow-up `remove()` runs, and the option is deleted — even though the caller had no ownership of it. The fix is to check `ownerUid in owners` before deciding to abort: if the caller is not in the map, return `owners` unchanged; only abort when `ownerUid` is present and is the sole entry.

- **`null` return commits the RTDB transaction instead of aborting it**: Firebase RTDB's transaction update function only aborts when `undefined` is returned — confirmed in the SDK source at line 11312 (`if (newVal === undefined)`). Returning `null` falls through to the commit branch and writes a null (deletion) to `ownerIds` with `result.committed = true`, so the `!result.committed` guard never fires and the option node is never deleted in the orphaned-ownerIds case. The `if (!owners) return owners` line should be `if (!owners) return undefined` to abort the transaction and let the follow-up `remove()` clean up the option.

- **Test for null-owners case asserts the wrong return value**: The test at line 95 (`expect(capturedUpdate?.(null)).toBeNull()`) asserts that the update function returns `null` for a null snapshot. But returning `null` commits the transaction rather than aborting it (see above), so this test is verifying incorrect behavior. Once the update function is fixed to return `undefined` when `owners` is null, this assertion must be updated to `toBeUndefined()`.

- **Missing test: non-owner calling unjoinOption should be a no-op**: There is no test covering the scenario where `ownerUid` is not present in `ownerIds` (e.g., `owners = { "user-2": true }` but `ownerUid = "user-1"`). Once the abort guard is fixed to check `ownerUid in owners`, a test should be added verifying that the update function returns `owners` unchanged and that `remove()` is not called.

## Verdict

Four correctness bugs confirmed in the diff — two in the transaction update logic and two in the tests. The null-return issue means the orphaned-option cleanup path is currently dead code (the transaction commits instead of aborting, so the follow-up delete never runs), and the missing ownership check allows any uid to delete an option they do not own.

## Deferred

The same `if (!members) return members` pattern exists in `removeMember` in `src/server/data/groups.ts` (introduced by PR #129, pre-existing in this PR). That function has the same null-return commit behavior and the same missing ownership check. Filed as a follow-up issue to track separately.

---
*Claude Sonnet 4.6*

**Comment by rmartz** [comment]
Follow-up on the Deferred note above: the same bugs in `removeMember` are tracked in [#175](https://github.com/rmartz/group-picks/issues/175).

---
*Posted by Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **`null` return commits the RTDB transaction instead of aborting it**: Fixed — `if (!owners) return owners` is now `if (!owners) return undefined`, correctly aborting the transaction for the null-snapshot case.
- ✅ **Non-owner uid can trigger option deletion (privilege escalation bug)**: Fixed — `if (!(ownerUid in owners)) return owners` guard added before the length check, returning `owners` unchanged when the caller is not a member of the owners map.
- ✅ **Test for null-owners case asserts the wrong return value**: Fixed — assertion updated from `toBeNull()` to `toBeUndefined()`, matching the corrected behavior.
- ✅ **Missing test: non-owner calling unjoinOption should be a no-op**: Fixed — new test added verifying `capturedUpdate?.(owners)` returns the same `owners` reference when the calling uid is absent from the map.

## Verdict

All four prior items resolved, incremental diff is clean, CI passing — ready to merge after UAT.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **`null` return commits the RTDB transaction instead of aborting it**: Fixed — `if (!owners) return owners` is now `if (!owners) return undefined`, correctly aborting the transaction for the null-snapshot case.
- ✅ **Non-owner uid can trigger option deletion (privilege escalation bug)**: Fixed — `if (!(ownerUid in owners)) return owners` guard added before the length check, returning `owners` unchanged when the caller is not a member of the owners map.
- ✅ **Test for null-owners case asserts the wrong return value**: Fixed — assertion updated from `toBeNull()` to `toBeUndefined()`, matching the corrected behavior.
- ✅ **Missing test: non-owner calling unjoinOption should be a no-op**: Fixed — new test added verifying `capturedUpdate?.(owners)` returns the same `owners` reference when the calling uid is absent from the map.

## Requested changes

- **Empty `owners` map (`{}`) is not handled — ownerless option survives**: When `owners = {}` (an empty object — a possible legacy/corrupt state), the `!(ownerUid in owners)` guard fires and returns `owners` unchanged, committing an empty `ownerIds` map and keeping an ownerless option alive. The old read-then-write path computed `remaining = []` in this case and deleted the option, so this is a behavioral regression for that degenerate state. Fix: add a check for `Object.keys(owners).length === 0` before the `ownerUid in owners` guard and return `undefined` to abort and trigger the follow-up `remove()`.

- **Missing test: empty `ownerIds` map should trigger option deletion**: No test covers `owners = {}`. Add a test asserting that the transaction update function returns `undefined` for an empty owners map, and that `remove()` is called (the option is cleaned up).

## Verdict

Two new Copilot-raised issues remain unaddressed — the empty `ownerIds` map edge case is a real behavioral regression from the old implementation and needs a small fix plus a corresponding test.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Empty `owners` map (`{}`) is not handled — ownerless option survives**: Fixed — `if (Object.keys(owners).length === 0) return undefined;` added before the `ownerUid in owners` guard, so an empty map now aborts the transaction and triggers the follow-up `remove()`, matching the old implementation's behavior.
- ✅ **Missing test: empty `ownerIds` map should trigger option deletion**: Fixed — test added asserting `capturedUpdate?.({})` returns `undefined`.

## Verdict

All prior items resolved and no new issues in the incremental diff — ready to merge after UAT.

---
*Claude Sonnet 4.6*

File: src/server/data/options.ts [hunk 1]
```diff
@@ -51,20 +51,29 @@ export async function unjoinOption(
...
+  // unjoin calls on a two-owner option can both see the other owner still
+  // present, both decide NOT to delete, and both remove only their own uid
+  // — leaving the option alive with an empty ownerIds map.
+  const result = await db
+    .ref(`picks/${pickId}/options/${optionId}/ownerIds`)
+    .transaction((owners: Record<string, true> | null) => {
+      if (!owners) return undefined;
+      if (Object.keys(owners).length === 0) return undefined;
+      if (!(ownerUid in owners)) return owners;
+      if (Object.keys(owners).length <= 1) return undefined;
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/server/data/options.ts:66]: Dismissing this. The two `Object.keys` calls serve logically distinct purposes — "is the map empty?" vs "is the calling uid the sole entry?" — and are separated by an early-return guard (`if (!(ownerUid in owners)) return owners`). Extracting a shared variable would not simplify the branching at this scale and the extra key enumeration is negligible.

— *Claude Sonnet 4.6*

File: src/server/data/options.spec.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        (update: (owners: Record<string, true> | null) => unknown) => {
+          capturedUpdate = update;
+          return Promise.resolve({ committed: true });
+        },
+      );
+
+      await unjoinOption("pick-1", "option-1", "user-1");
+
+      const owners = { "user-2": true as const };
+      expect(capturedUpdate?.(owners)).toBe(owners);
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/server/data/options.spec.ts:127]: Dismissing this. The implementation explicitly returns the same `owners` reference (`return owners`) to signal "nothing changed here, pass through unchanged." Using `toBe` rather than `toEqual` is intentional: it verifies that the update function returns the original object identity (no unnecessary copy), which is the semantically correct property to assert on the pass-through path.

— *Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Empty `owners` map (`{}`) is not handled — ownerless option survives**: Fixed — `if (Object.keys(owners).length === 0) return undefined;` added before the `ownerUid in owners` guard.
- ✅ **Missing test: empty `ownerIds` map should trigger option deletion**: Fixed — test added asserting `capturedUpdate?.({})` returns `undefined`.
- ✅ [src/server/data/options.ts line 66 (copilot-pull-request-reviewer)](https://github.com/rmartz/group-picks/pull/153#discussion_r3243477168): Explicitly dismissed — the two `Object.keys` calls serve distinct purposes and are separated by an early-return guard.
- ✅ [src/server/data/options.spec.ts line 127 (copilot-pull-request-reviewer)](https://github.com/rmartz/group-picks/pull/153#discussion_r3243477221): Explicitly dismissed — `toBe` is intentional; the update function returns the original reference to signal "pass through unchanged," and asserting object identity is the semantically correct check here.

## Requested changes

- **PR description states "9 tests added" but the spec file contains 11 `it(...)` cases**: The test count in the description became stale across review cycles — the empty-map and non-owner test cases added in later commits bring the total to 11. Update the description to reflect the actual count so it stays in sync with the code per CLAUDE.md's documentation rule.

## Verdict

One minor documentation inaccuracy to fix: the PR description test count is stale. One inline thread (test count) remains open.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **PR description states "9 tests added" but the spec file contains 11 `it(...)` cases**: Fixed — description now correctly reads "11 tests added, all passing."

## Verdict

All prior items resolved and no new issues in the incremental diff (the latest commit is a branch-update merge from main with no changes to the PR's own files) — ready to merge after UAT.

---
*Claude Sonnet 4.6*

**Comment by rmartz** [comment]
**Pre-merge check**

- ✅ No WIP or draft
- ✅ PR approved, no changes requested
- ✅ UAT complete (skip-uat active)
- ✅ Branch up to date, no conflicts
- ✅ All CI green

Merging.

---
*Posted by Claude Sonnet 4.6*
